File: test/python/timer.py
from threading import Timer
import datetime

timeout = 2000
simultitaneous = 100
results = []

def timeFinished(arg1, arg2):
    global results
    #results.append(datetime.datetime.now() - arg1 - timeout/1000.0)
    results.append((datetime.datetime.now() - arg1 - datetime.timedelta(milliseconds=timeout)).total_seconds())
    k = (sum(results) / float(len(results))*1000)

def allFinished():
    global results
    print 'python\t|timer\t| Average delay between expected and real timeout is ' + str((sum(results) / float(len(results))*1000)) + 'ms'
    #print (sum(results) / float(len(results))*1000)

#arguments: 
#how long to wait (in seconds), 
#what function to call, 
#what gets passed in
for i in range(0, simultitaneous):
  r = Timer(timeout/1000.0, timeFinished, (datetime.datetime.now(), "arg2"))
  r.start()
s = Timer(timeout/1000.0+ 0.5, allFinished)

s.start()

File: test/cpp/timer.cpp

#include <iostream>
#include <thread>
#include <functional>
#include <numeric>
#include <chrono>


void delayer(std::function<void()> f, int ms)
{
    std::this_thread::sleep_for(std::chrono::microseconds(ms));
    f();
}

int main()
{
    constexpr int expectedTimeout = 2000;
    constexpr int simultaneous = 100;

    int results[simultaneous];
    std::thread threadPool[simultaneous];

    for (int i = 0; i < simultaneous; ++i)
    {
        auto beforeDelay = std::chrono::system_clock::now();

        auto f = [&results, i, beforeDelay]()
        {
            auto actualTimeout = std::chrono::system_clock::now() - beforeDelay;
            results[i] = actualTimeout.count() / 1000;
        };

        threadPool[i] = std::thread(delayer, f, expectedTimeout);
    }

    for (std::thread& thread : threadPool)
    {
        thread.join();
    }

    int average = std::accumulate(results, results + simultaneous, 0) / simultaneous;
    std::cout << "cpp\t|timer\t| Average delay between expected and real timeout is " << average - expectedTimeout << "ms" << std::endl;

    return 0;
}
